sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -28,6 +30,34 @@ import org.apache.spark.unsafe.Platform
 import org.apache.spark.unsafe.array.ByteArrayMethods
 import org.apache.spark.unsafe.types.{ByteArray, UTF8String}
 
+/**
+ * Base trait for [[BinaryExpression]]s with two arrays of the same element type and implicit
+ * casting.
+ */
+trait BinaryArrayExpressionWithImplicitCast extends BinaryExpression
+  with ImplicitCastInputTypes {
+
+  protected lazy val elementType: DataType = inputTypes.head.asInstanceOf[ArrayType].elementType
+
+  override def inputTypes: Seq[AbstractDataType] = {
+    TypeCoercion.findWiderTypeForTwo(left.dataType, right.dataType) match {
```

Comment:
does presto allow implicitly casting to string for these collection functions? e.g. can `ArraysOverlap` work for array of int and array of string?

Reply:
it does not for int and string, but it does for decimal and int. What shall we do?

Reply:
Then we probably need to call `TypeCoercion.findTightestCommonType` here, and fix `findTightestCommonType` for array/struct/map types.

Reply:
what about `findWiderTypeWithoutStringPromotionForTwo`?

Reply:
now the question is, shall we allow precision lose for array functions?

Reply:
good question. Checking the way we are doing it I would say no. Since we are bounding in a quite strange way at the moment (causing loss of int digits instead of decimals) I would say no, since this could lead to have many `NULL`s. What do you think?

Reply:
yes, I'll go for it. Thanks.